versioned_hdf5/replay.py

```diff
@@ -316,9 +347,143 @@ def _recreate_virtual_dataset(f, name, versions, raw_data_chunks_map, tmp=False)
             del group[name]
             group.move(tmp_path, name)
 
-def delete_versions(f, versions_to_delete):
+
+def _is_empty(f: VersionedHDF5File, name: str, version: str) -> bool:
+    """Return True if the dataset at the given version is empty, False otherwise.
+
+    Assumes the dataset exists in the given verison.
+
+    Parameters
+    ----------
+    f : VersionedHDF5File
+        File where the dataset resides
+    name : str
+        Name of the dataset
+    version : str
+        Version of the dataset to check
+
+    Returns
+    -------
+    bool
+        True if the dataset is empty, False otherwise
     """
-    Completely delete the versions from `versions_to_delete` from the versioned file `f`.
+    return not f['_version_data/versions'][version][name].is_virtual
+
+
+def _exists_in_version(f: VersionedHDF5File, name: str, version: str) -> bool:
+    """Check if a dataset exists in a given version.
+
+    Parameters
+    ----------
+    f : VersionedHDF5File
+        File where the dataset may reside
+    name : str
+        Name of the dataset
+    version : str
+        Version of the dataset to check
+
+    Returns
+    -------
+    bool
+        True if the dataset exists in the version, False otherwise
+    """
+    return name in f['_version_data/versions'][version]
+
+
+def _all_extant_are_empty(
+    f: VersionedHDF5File, name: str, versions: Iterable[str]
+) -> bool:
+    """Check if the given versions of a dataset are empty.
+
+    Doesn't assume the dataset exists in any version.
+
+    Parameters
+    ----------
+    f : VersionedHDF5File
+        File where the dataset may reside
+    name : str
+        Name of the dataset
+    version : str
+        Version of the dataset to check
+
+    Returns
+    -------
+    bool
+        True if any version of the dataset that can be found is empty,
+        False if a version exists which is not.
+    """
+    for version in versions:
+        if _exists_in_version(f, name, version):
+            if not _is_empty(f, name, version):
+                return False
+    return True
+
+
+def _delete_dataset(f: VersionedHDF5File, name: str, versions_to_delete: Iterable[str]):
+    """Delete the given dataset from the versions."""
+    version_data = f['_version_data']
+    versions = version_data['versions']
+
+    if name == 'versions':
+        return
+
+    versions_to_keep = set(versions) - set(versions_to_delete)
+
+    # If the dataset is empty in the versions to delete, we don't
+    # need to recreate the raw data, hash table, or virtual datasets.
+    if _all_extant_are_empty(f, name, versions_to_delete):
```

Comment:
Is this is the only effective change here?

Reply:
Yep.